Update logging guidelines (#6283)
kubernetes/community: contributors/devel/sig-instrumentation/logging.md

## Logging

> This document will provide an overview on how Kubernetes community recommendation on logging. 

Reviewer: ```suggestion
This document provides an overview of  the recommended way to develop and implement
logging for components of Kubernetes.
```

> Following recommendation should be followed in https://github.com/kubernetes/kubernetes, however any Kubernetes or external project is fill to follow guidelines below as most of the ideas can be applied more broadly.

Reviewer: ```suggestion
For [k/kubernetes](https://github.com/kubernetes/kubernetes) these conventions should be seen
as a strong recommendation, however any Kubernetes or external project is welcome to follow
the guidelines below. Most of the ideas can be applied more broadly.
```

### Logging in Kubernetes

Kubernetes project uses [klog](https://github.com/kubernetes/klog) for logging.
This library was created as a permanent fork of glog, as the original project was no longer developed. 
However, should not be used by external projects as it is mostly developed to maintain backward compatibility.

> With introduction of [Structured Logging](https://github.com/kubernetes/enhancements/issues/1602) Kubernetes is being migrated to [logr](https://github.com/go-logr/logr) interface, any changes in klog are done to ensure smooth transition.

Reviewer: ```suggestion
With the introduction of [Structured Logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging), Kubernetes is being migrated to use [logr](https://github.com/go-logr/logr) as its logging interface. Any changes in klog are done to ensure a smooth transition to `logr`.
``

> For projects that want to integrate with Kubernetes or write same log format we recommend using [klogr](https://github.com/kubernetes/klog/tree/main/klogr), reimplementation of klog format that exposes [logr](https://github.com/go-logr/logr) interface.

Reviewer: Is this correct?
```suggestion
For projects that want to integrate with Kubernetes or write same log format we recommend using [klogr](https://github.com/kubernetes/klog/tree/main/klogr). The `klogr` library exposes an interface that
follows [logr](https://github.com/go-logr/logr) conventions, and emits log entries that are equivalent to `klog`.
```

Reviewer 2: I don't think we should recommend using klogr, or at least not without further explanations. For example, a library should accept a `logr.Logger` instance and let the caller decide whether that is klogr or something else.

But right now we are not ready yet to propose *how* such a library should accept such a logger - we need a decision about https://github.com/kubernetes/enhancements/pull/3078 first.

### Logging Conventions

### When not to log?

> Shared libraries, such as `client-go`, should not log write logs themselves,

Reviewer: ```suggestion
Shared libraries, such as [`client-go`](https://github.com/kubernetes/client-go), should _not_
write logs themselves,
```

Reviewer 2: A language question: what does "logs" stand for here and in other places ("Logs flags" in kube-apiserver)?

Is it short for "log files" or "log messages"?

"to log" is a verb, but here it is uses as a noun. I would use "logging" when I need a noun for the activity or "log messages" when I need an object ("write log messages").

Reviewer: Historically the log was a noun. Originally it referred to a flotation device attached to a ship's waterproofed record book, but it has since come to mean the journal itself, or more commonly any broadly equivalent set of chronological, time-stampable records.
Further language drift added the verb “to log”. That use is fine too but it is actually a subsidiary meaning.

Reviewer 2: >it has since come to mean the journal itself, or more commonly any broadly equivalent set of chronological, time-stampable records

So there is *one* of those. Does it then make sense to use the plural form or is singular better?

I understand "log books" (nautical), "log files" (computer science), but "logs" (plural for an abstract concept in computer science?) still sounds odd to me.

Reviewer 2: Then I prefer "log messages" instead of the ambiguous "logs" abbreviation. I'm using "messages" instead of "entries" because "message" seems more common in computer science.

Author: I always interpreted "log message" as part of whole log that was provided by developer, so a log minus metadata like timestamp, caller etc.

Maybe we should also write down naming conventions for logging that all documentation should follow.

but just return `error`, because client libraries may be used in CLI UIs that wish to control output.

### How to log

> There are two main klog methods for writing logs `klog.InfoS` and `klog.ErrorS` as they provide [logr](https://github.com/go-logr/logr) compatible interface.

Reviewer: ```suggestion
There are two main klog methods for writing logs: `klog.InfoS` and `klog.ErrorS`. Both methods are part of a [logr](https://github.com/go-logr/logr) compatible interface.
```

> For those two functions exists additional flavors, like `klog.V(2).InfoS` that allows caller to increase log verbosity (making it only available when `--v=2` flag is provided to binary) and `klog.InfoSDepth` with `klog.ErorSDepth` that give control to change caller information.

Reviewer: Is this accurate?
```suggestion
You can also find variants of those functions, such as `klog.V(2).InfoS` that allows caller to increase log verbosity (making it only available when `--v=2` flag is provided toa  binary). Another example is `klog.InfoSDepth` and `klog.ErorSDepth` that provide a way to report caller information.
```

Author: Part about caller information is wrong. Caller information is always provided, it points to code location that called log library. `klog.InfoSDepth` is meant to give some control over caller information to author of code.

Calling just `Info` works in cases where logging library is called directly, however in some cases it's useful to create logging utility functions that log themselves removing repetition or ensuring consistency. In such cases log library is called indirectly, resulting in caller information pointing to utility function. This becomes a problem when utility function is used frequently resulting caller information becoming useless as all different calls point to same place.

Any other non-structured (without `S`) klog methods like `klog.Infof` that use old C like string format interface are no longer recommended for use. 
Those log calls using those methods should be migrated to their structured variants.
Please see the [Structured Logging Guide](migration-to-structured-logging.md#structured-logging-in-kubernetes) for more information on how to migrate such logs.

> All structured method accept log message as string and any number of key value pairs provided via variadic `interface{}` argument.

Reviewer: ```suggestion
All structured logging methods accept the log message as a string, along with any number of key/value pairs that you provide via a variadic `interface{}` argument.
```

Prototype:
```go
func InfoS(message string, keysAndValues ...interface{})
func InfoSDepth(depth int, message string, keysAndValues ...interface{})
func ErrorS(err error, message string, keysAndValues ...interface{})
func ErrorSDepth(depth int, err error, message string, keysAndValues ...interface{})
```

Example:
```go
klog.InfoS("Received HTTP request", "method", "GET", "URL", "/metrics", "latency", time.Second)
```

### What method to use?

* `klog.ErrorS` - Errors should be used to indicate unexpected behaviours in code, like unexpected errors returned by subroutine function calls.
Logs generated by `ErrorS` command may be enhanced with additional debug information (by logging library). Calling `ErrorS` with `nil` as error may be acceptable if there is error condition that deserves a stack trace at this origin point.

* `klog.InfoS` -  Structured logs to the INFO log. `InfoS` should be used for routine logging. It can also be used to log warnings for expected errors (errors that can happen during routine operations).
  Depending on log severity it's important to pick a proper verbosity level to ensure that consumer is neither under nor overwhelmed by log volume:

>   * `klog.V(0).InfoS` - Generally useful for this to ALWAYS be visible to an operator

Reviewer: ```suggestion
  * `klog.V(0).InfoS` - Generally useful for this to **always** be visible to a cluster operator
```

logging for components of Kubernetes.
For [k/kubernetes](https://github.com/kubernetes/kubernetes) these conventions should be seen
as a strong recommendation, however any Kubernetes or external project is welcome to follow
the guidelines below. Most of the ideas can be applied more broadly.

### Logging in Kubernetes

Kubernetes project uses [klog](https://github.com/kubernetes/klog) for logging.
This library was created as a permanent fork of glog, as the original project was no longer developed. 
With the introduction of [Structured Logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging) and [Contextual Logging](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/3077-contextual-logging/README.md), Kubernetes is being migrated to use [logr](https://github.com/go-logr/logr) as its logging interface. Any changes in klog are done to ensure a smooth transition to `logr`. Klog is then only used to manage the logging configuration (`SetLogger`, retrieving the logger) and its logging calls should not be used anymore.
``

> For projects that want to integrate with Kubernetes or write the same log format, retrieving a `logr.Logger` instance from klog will return an implementation that emits log output the same was the traditional klog logging calls.

Reviewer 2: Sorry, I had a typo in my original suggestion:

```suggestion
For projects that want to integrate with Kubernetes or write the same log format, retrieving a `logr.Logger` instance from klog will return an implementation that emits log output the same way as the traditional klog logging calls.
```

For [k/kubernetes](https://github.com/kubernetes/kubernetes) these conventions should be seen
as a strong recommendation, however any Kubernetes or external project is welcome to follow
the guidelines below. Most of the ideas can be applied more broadly.

### Logging in Kubernetes

Kubernetes project uses [klog](https://github.com/kubernetes/klog) for logging.
This library was created as a permanent fork of glog, as the original project was no longer developed. 
However, should not be used by external projects as it is mostly developed to maintain backward compatibility.
With the introduction of [Structured Logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging), Kubernetes is being migrated to use [logr](https://github.com/go-logr/logr) as its logging interface. Any changes in klog are done to ensure a smooth transition to `logr`.
``

For projects that want to integrate with Kubernetes or write same log format we recommend using [klogr](https://github.com/kubernetes/klog/tree/main/klogr), reimplementation of klog format that exposes [logr](https://github.com/go-logr/logr) interface.

### Logging Conventions

> ### When not to log?
> 
> Shared libraries, such as [`client-go`](https://github.com/kubernetes/client-go), should _not_
> write logs themselves,
> but just return `error`, because client libraries may be used in CLI UIs that wish to control output.

Reviewer 2: ```suggestion
### When not to log?

Shared libraries, such as [`client-go`](https://github.com/kubernetes/client-go), should _not_
log errors themselves
but just return `error`, because client libraries may be used in CLI UIs that wish to control output.
```

### How to log

> There are two main klog methods for writing logs: `klog.InfoS` and `klog.ErrorS`. Both methods are part of a [logr](https://github.com/go-logr/logr) compatible interface.

Reviewer 2: I will have to update this section to cover the replacement of those calls with logging through `Logger`. This goes beyond a simple PR review, so let's get this merged first even if it is not fully up-to-date.

Author: sg

Prototype:
```go
func InfoS(message string, keysAndValues ...interface{})
func InfoSDepth(depth int, message string, keysAndValues ...interface{})
func ErrorS(err error, message string, keysAndValues ...interface{})
func ErrorSDepth(depth int, err error, message string, keysAndValues ...interface{})
```

Example:
```go
klog.InfoS("Received HTTP request", "method", "GET", "URL", "/metrics", "latency", time.Second)
```

### What method to use?

* `klog.ErrorS` - Errors should be used to indicate unexpected behaviours in code, like unexpected errors returned by subroutine function calls.
Logs generated by `ErrorS` command may be enhanced with additional debug information (by logging library). Calling `ErrorS` with `nil` as error may be acceptable if there is error condition that deserves a stack trace at this origin point.

* `klog.InfoS` -  Structured logs to the INFO log. `InfoS` should be used for routine logging. It can also be used to log warnings for expected errors (errors that can happen during routine operations).
  Depending on log severity it's important to pick a proper verbosity level to ensure that consumer is neither under nor overwhelmed by log volume:

>   * `klog.V(0).InfoS` - Generally useful for this to **always** be visible to a cluster operator

Reviewer 2: ```suggestion
  * `klog.V(0).InfoS` = `klog.InfoS` - Generally useful for this to **always** be visible to a cluster operator
```
